### PR TITLE
Feat/backend worldmap loader

### PR DESCRIPTION
**Pull request by ilyasfoo**
Fixes https://github.com/Code-for-All/lockdown/issues/100 and https://github.com/Code-for-All/lockdown/issues/103

### CODE REVIEW

**Comment by LarsDenBakker** [comment]
This will also require some front-end changes, right @thepassle ?

**Comment by thepassle** [comment]
Could you add the new output JSON files to this PR so I can check the schema?

@LarsDenBakker this will probably need frontend changes yeah, but the colouring of the worldmap was already broken since the last changes. The fixes for https://github.com/Code-for-All/lockdown/issues/100 need changes, and the fixes for https://github.com/Code-for-All/lockdown/issues/103 hasnt been implemented yet (country details dialog)

**Comment by ilyasfoo** [comment]
Ok i've added to this PR, please check @thepassle

File: backend/babel.config.js [hunk 1]
```diff
@@ -8,6 +8,10 @@ module.exports = {
       }
     ]
   ],
+  "plugins": [
+    "@babel/plugin-proposal-nullish-coalescing-operator",
```
Comment (thepassle) [inline backend/babel.config.js]: Were these intentionally added? You dont seem to use nullish coalescing or optional chaining in the backend code yet, or do you intend to in the future?

**Comment by thepassle** [comment]
So these will have all the 'entries' combined/merged to one 'current' status per country, correct?

These look excellent:

```json
{
   "lockdown":{
      "editor":"Irene",
      "reviewed_by":"",
      "status":"Ready",
      "type":"Proclamation",
      "date_of_entry":"1 March 2020",
      "name":"Whatever 1",
      "url":"www.ua.es",
      "title":"Other (Yes)",
      "date":"5 March 2020",
      "measures":[
         {
            "label":"max_gathering",
            "value":"1000"
         },
         {
            "label":"lockdown_status",
            "value":"No",
            "start":"5 March 2020",
            "end":"15 March 2020"
         },
         {
            "label":"city_movement_restriction",
            "value":"Yes",
            "start":"5 March 2020",
            "end":"15 March 2020"
         },
         {
            "label":"attending_religious_sites",
            "value":"Yes",
            "start":"5 March 2020",
            "end":"15 March 2020"
         },
         {
            "label":"going_to_work",
            "value":null
         },
         {
            "label":"military_not_deployed",
            "value":"Yes",
            "start":"5 March 2020",
            "end":"15 March 2020"
         },
         {
            "label":"academia_allowed",
            "value":null
         },
         {
            "label":"going_to_shops",
            "value":"Yes",
            "start":"5 March 2020",
            "end":"10 March 2020"
         },
         {
            "label":"electricity_nominal",
            "value":null
         },
         {
            "label":"water_nominal",
            "value":null
         },
         {
            "label":"internet_nominal",
            "value":null
         }
      ],
      "travel":{
         "land":[
            {
               "label":"local",
               "value":null
            },
            {
               "label":"nationals_inbound",
               "value":null
            },
            {
               "label":"nationals_outbound",
               "value":null
            },
            {
               "label":"foreigners_inbound",
               "value":null
            },
            {
               "label":"foreigners_outbound",
               "value":null
            },
            {
               "label":"cross_border_workers",
               "value":null
            },
            {
               "label":"commerce",
               "value":null
            }
         ],
         "flight":[
            {
               "label":"local",
               "value":null
            },
            {
               "label":"nationals_inbound",
               "value":null
            },
            {
               "label":"nationals_outbound",
               "value":null
            },
            {
               "label":"foreigners_inbound",
               "value":null
            },
            {
               "label":"foreigners_outbound",
               "value":null
            },
            {
               "label":"stopovers",
               "value":null
            },
            {
               "label":"commerce",
               "value":null
            }
         ],
         "sea":[
            {
               "label":"local",
               "value":null
            },
            {
               "label":"nationals_inbound",
               "value":null
            },
            {
               "label":"nationals_outbound",
               "value":null
            },
            {
               "label":"foreigners_inbound",
               "value":null
            },
            {
               "label":"foreigners_outbound",
               "value":null
            },
            {
               "label":"cross_border_workers",
               "value":null
            },
            {
               "label":"commerce",
               "value":null
            }
         ]
      }
   }
}
```

**Comment by thepassle** [comment]
These look great as well :) Nice work!

LGTM, @LarsDenBakker any more comments?

```js
{
      "type": "Feature",
      "properties": {
        "iso2": "AF",
        "iso3": "AFG",
        "NAME": "Afghanistan",
        "lockdown_status": "No"
      },
      "geometry": {/* */}
    },
```

**Comment by thepassle** [comment]
One more question, this data is fetched from the `playground` tab in google sheets, correct? Do you need @JFQueralt to add tabs per country again and then update the script? Because I think editors can already start entering data now, right?

File: backend/babel.config.js (hunk 1 above)
Comment (ilyasfoo) [reply backend/babel.config.js]: Yes I was working with enum in a different branch but added these 2 plugins together

**Comment by ilyasfoo** [comment]
> So these will have all the 'entries' combined/merged to one 'current' status per country, correct?

It's not merged, the assumption is the first entry per country is the "current" status

> One more question, this data is fetched from the playground tab in google sheets, correct? Do you need @JFQueralt to add tabs per country again and then update the script? Because I think editors can already start entering data now, right?

Nope, it's set to [DEMO](https://docs.google.com/spreadsheets/d/1mVyQxxLxAF3E1dw870WHXTOLgYzmumojvzIekpgvLV0/edit#gid=1961129897), had more data than Playground IMO

**Comment by thepassle** [comment]
> It's not merged, the assumption is the first entry per country is the "current" status

Are we sure this assumption is correct? @JFQueralt can you confirm?

> Nope, it's set to DEMO, had more data than Playground IMO. Yes will need the tabs per country & update script

Okay, clear. Thank you :)

So I think this means @JFQueralt can start making the tabs for every country, and editors can start submitted data, right?

**Comment by JFQueralt** [comment]
> > It's not merged, the assumption is the first entry per country is the "current" status
> 
> Are we sure this assumption is correct? @JFQueralt can you confirm?

I am developing the concept for that. Had several conversations about the most convenient procedure. Still working my head on it.

> > Nope, it's set to DEMO, had more data than Playground IMO. Yes will need the tabs per country & update script
> 
> Okay, clear. Thank you :)
> 
> So I think this means @JFQueralt can start making the tabs for every country, and editors can start submitted data, right?

Yes, I hope to finish the Template tab soon. I realized that some script intensively calls Google's servers and that's no good for impact and I suspect for API quota. Re-working those functions.

I should be able to create all the tabs soon ALTHOUGH that should not affect you anymore, truly.
What you guys need to extract now are the XXX_Fetch Tabs, which are the tabs with the ALREADY processed values for each Data Point on a "calendar view". I am calling them Snapshots.
All the business logic is taken care of by the G Sheet.
So far Ilyas has a sample tab "AFG_Fetch" he can use to extract and you can build the client-side.

Happy to have a call if that will help. I attach a diagram that should be self-explanatory.


photo_2020-03-31_18-09-00

**Comment by thepassle** [comment]
@JFQueralt I was thinking it would look like this:

image

**Comment by JFQueralt** [comment]
> 
> 
> @JFQueralt I was thinking it would look like this:
> 
> image


The problem is that the order of entering does not need to be chronological AND some pronounciations override/clash with previous ones. For instance:

- Doc1, issued on 25th Feb:
*Max people 1000 from 1st March to 31st March
*Stay home from 1st March to 15th March

then 5 days later they go "holy... we were wrong!" then they issue:

- Doc2, issued on 5th March 
*Max people 500 from 6th March to 31st March

2 things happen there:
1) Dates overlap so the business logic needs to take that into account
2) Editors may know about Doc2 but not Doc1... and they find Doc1 on the 10th March.

You'd still need to apply the docs' mandates yet they wouldn't necessarily be in the "right chronological order". You can never be sure of that. Nor that ever doc will establish "a starting status" for each Data Point. The reality is that docs are issed in an incremental way, only mentioning whay they think they to update.

And this is why it's so critical that we show somewhere (next iteration) the log of which docs are used to reach to a certain conclusion.

You can imagine the clusterfuck of pushing this business logic on the client. Hence the snapshots: they reflect the state of each Data Point considering the Data Entries we have.

This will also explain why we are using Updates: To broadcast when we discover "something wasn't right because wee didn't know about this particular Doc". Corrections are a normal thing; transparency will save us.

Jean

**Comment by thepassle** [comment]
Okay, I will let you guys figure it out 😅

**Comment by JFQueralt** [comment]
I thought we had the country data sources figured out. World Bank?

@milovanderlinden

**Comment by ilyasfoo** [comment]
@JFQueralt I think what @thepassle meant was can we confirm on how to obtain the current state per country.

The backend script currently takes the *first* entry of each country tab as current. If this is aligned with our V1 outcome, we can proceed with this PR, no change to code is needed

**Comment by thepassle** [comment]
> @JFQueralt I think what @thepassle meant was can we confirm on how to obtain the current state per country.

@JFQueralt  This is correct, imo this is the main blocker right now. If the first entry for a country is the most up to date status that we should display in the country detail dialog, we can move forward. If thats not the case, we need to talk about a solution/how to solve it

Additionally we need data for the 'ticker(updates) https://github.com/Code-for-All/lockdown/issues/79, totals - people affected https://github.com/Code-for-All/lockdown/issues/75, totals - countries in lockdown https://github.com/Code-for-All/lockdown/issues/74

But I would say lets first try to move forward with the current isssue here, which is getting the most up to date status per country for the country detail dialog

**Comment by thepassle** [comment]
As an extra comment: can we minify the json before we save the files? See https://github.com/Code-for-All/lockdown/issues/116

**Comment by ilyasfoo** [comment]
Just ended call with @JFQueralt , we've confirmed that the latest status in V1 is going to be determined by current date. Script will process this in UTC, and produce the latest data. One important change is that script will need to fetch from the new "snapshot" sheet.

**Comment by ilyasfoo** [comment]
Closing this, please refer to https://github.com/Code-for-All/lockdown/pull/129

**Pull request closed without merging**